File: 05ovo2e/Week8/게임 맵 최단거리.cpp
#include<vector>
#include <queue>
#include <tuple>

using namespace std;

// 가중치가 같은 그래프 내의 최단거리 알고리즘
// 한 칸씩 이동할 수 있으므로
// BFS로 최단거리 배열을 만든다.

const int max_n = 104;
int dy[4] = {-1, 0, 1, 0}; // 상하
int dx[4] = {0, 1, 0, -1}; // 좌우
int n, m, visited[max_n][max_n], y, x;

int solution(vector<vector<int>> maps)
{
    int answer = 0;
    
    n = maps.size();
    m = maps[0].size();
    
    queue<pair<int, int>> q;
    
    visited[0][0] = 1;
    q.push({0, 0});
    
    while (q.size()) {
        tie(y,x) = q.front();
        q.pop();
        for (int i=0; i<4; i++) {
            int ny = y + dy[i];
            int nx = x + dx[i];
            if(ny < 0 || ny >= n || nx < 0 || nx >= m || maps[ny][nx] == 0)
                continue;
            if(visited[ny][nx]) continue;
            visited[ny][nx] = visited[y][x] + 1;
            q.push({ny, nx});
        }
    }
    
    if (visited[n-1][m-1] == 0) {
        return -1;
    }

    return visited[n-1][m-1];
}

File: arim/Week8/게임 맵 최단거리.java
import java.util.*;

class Solution {
    int[] dx={0,1,-1,0};
    int[] dy={1,0,0,-1};
    
    public int solution(int[][] maps) {
        int answer = 0;
        int[][] visit=new int[maps.length][maps[0].length];
        
        bfs(maps, visit);
        answer=visit[maps.length-1][maps[0].length-1];
        
        if(answer==0) answer=-1;
        return answer;
    }
    
    public void bfs(int[][] maps, int[][] visit){
        Queue<int[]> q=new LinkedList<>();
        q.add(new int[]{0, 0});
        visit[0][0]=1;
        
        while(!q.isEmpty()){
            int[] cur=q.poll();
            int x=cur[0];
            int y=cur[1];
            
            for(int i=0; i<4; i++){
                int nx=x+dx[i];
                int ny=y+dy[i];
                
                if(nx<0||ny<0||nx>maps.length-1||ny>maps[0].length-1) continue;
                if(visit[nx][ny]==0&&maps[nx][ny]==1){
                    visit[nx][ny]=visit[x][y]+1;
                    q.add(new int[]{nx, ny});
                }
            }
        }
    }
}